backend/api/src/create-announcement.ts

```diff
@@ -0,0 +1,57 @@
+import * as admin from 'firebase-admin'
+import { z } from 'zod'
+import { APIError, authEndpoint, validate } from './helpers'
+import { createCustomNotification } from 'shared/create-notification'
+import { getUser } from 'shared/utils'
+import { User } from 'common/user'
+import {
+  SupabaseDirectClient,
+  createSupabaseDirectClient,
+} from 'shared/supabase/init'
+import { isAdminId } from 'common/envs/constants'
+
+async function getAllUsersPaginated(pg: SupabaseDirectClient) {
+  const pageSize = 1000
+  let offset = 0
+  let users: User[] = []
+
+  while (true) {
+    const chunk = await pg.map(
+      `select data from users limit $1 offset $2`,
+      [pageSize, offset],
+      (r: any) => r.data as User
+    )
+
+    if (chunk.length === 0) break
+
+    users = [...users, ...chunk]
+    offset += pageSize
+  }
+
+  return users
+}
+
+const bodySchema = z.object({
+  url: z.string(),
+  title: z.string(),
+})
+
+export const createannouncement = authEndpoint(async (req, auth) => {
+  const pg = createSupabaseDirectClient()
+  const isAdmin = isAdminId(auth.uid)
+  const { title, url } = validate(bodySchema, req.body)
+
+  if (!isAdmin) throw new APIError(403, 'Insufficient permissions.')
```

Comment:
you can put this error before the validation, even.

Reply:
And put the isAdmin code before the `const pg = createSupabaseDirectClient()` line